Fixed bugs to support mixture of sunsynk inverter plant lists

t0mtaylor opened the pull request:
Made some improvements which should help first timers with the scripts, tested on latest HA 2024 and Node Red v3

- Multiple lists of plants can have master and slaves and different size inverters, which broke part of the flow

- Time uploads broke as didn't follow the correct format - mainly due to slave plants trying to send zero

- This config has been tested with the latest Docker containers of HA 2024 and NR v3

- Added additional debugging for debug mode

- Display the debug and warnings state is enabled/disabled within Flow and debug msgs

- Performance improvements, which skip Plant ids where an inverter id and data doesn't exist (usually slave plants) - this helps with multi-plant accounts, also where an inactive plant may exist within the list also

- ~~Removed region 1 (as we should now be using region 2) and inteless.com references (Power View)~~

- ~~Updated config to use api.sunsynk.net and sunsynk.net for login, etc~~

- Tidy up debug messages, add friendly Warnings to help with setup issues for first setup, as current wiki not clear

- ~~Reorder template as this broke in newer versions of HA 2024~~

link to post - https://powerforum.co.za/topic/16806-sunsynk-logger-integration-into-home-assistant/?do=findComment&comment=195404

t0mtaylor committed:
Fixed bugs to support inverter modes

See PR for more details

t0mtaylor commented on V2.3/JSON - Node Red - Flows Required.json:
```diff
@@ -1,16 +1,45 @@
+        "type": "server",
+        "name": "Home Assistant",
+        "version": 5,
+        "addon": false,
```
Changing key 'addon' to false will work with separate docker install of HA and NR, just remember to set the private key from HA user, and use external IP with port map for hass-config if accessing the container directly does not work

gdwaterworth replied:
This still need to look at. I prefer the orrigional way for a few reasons, but am considering this

t0mtaylor committed:
Tidy up debug, add friendly Warnings to help with setup issues

t0mtaylor committed:
reorder template as this broke in newer versions of HA 2024

t0mtaylor committed:
Show if debug warnings are enabled within Flow

gdwaterworth commented:
- Removed region 1 (as we should now be using region 2) and inteless.com references (Power View)
This is incorrect as this is also usable by other systems that are still on region 1

- Time uploads broke as didn't follow the correct format
 More Information please as the format is forced via teh input_select

- Performance improvements, which skip Plant ids where an inverter id and data doesn't exist (usually slave plants)
Will think on this as I am not sold on filtering the raw data 
Besides if there are no inverters , the calls will not be made to get inverter data 
So unneeded coding 

- Updated config to use api.sunsynk.net and sunsynk.net for login, etc
Unneeded

- Reorder template as this broke in newer versions of HA 2024
More information please. The latest version is running fine on the latest HA and Nodered

- Multiple lists of plants can have master and slaves and different size inverters, which broke part of the flow
Raw data gather should be fine on this, so more info please
Export flow doesnt currently cater for multi plant

t0mtaylor committed:
Resolved comments from PR

t0mtaylor committed:
Resolved issue with new template declaration at top of file

t0mtaylor commented:
1. TT: Removed region 1 (as we should now be using region 2) and inteless.com references (Power View)
GW: This is incorrect as this is also usable by other systems that are still on region 1
**TT: I'll re-add Region 1, as my focus was on Region 2**

2. TT: Time uploads broke as didn't follow the correct format
GW: More Information please as the format is forced via teh input_select
**TT: The input_select didn't seem to work as expected, the new time wrapper function seemed to fix the issue where the time was not stored correctly as it was default to 0 zero when the inverter does not provide the information (could be linked to issue 3 below), so this function is needed for certain inverters, as the data is not expected with the correct time format (due to zero)**

3. TT: Performance improvements, which skip Plant ids where an inverter id and data doesn't exist (usually slave plants)
GW: Will think on this as I am not sold on filtering the raw data
GW: Besides if there are no inverters , the calls will not be made to get inverter data
GW: So unneeded coding
**TT: The fixes are within `Export Data Export To Entity` - raw data your trying to filter does not existing for slave inverters but it is ok on master/single inverters - these slave inverters do exist within the list but have missing data you expected (and that is the issue) when a slave inverter is linked to a master inverter (on the account we have 3 installs which has 5 inverters, 2 installs are master and slave 5.5kw and the 3rd is a single 8.8kw), as the `inverter` key value object does not exist on slaves - you are then looping through the values within `Export Data Export To Entity` which don't exist, which seems pretty pointless. This also caused a javascript error as the object was undefined on the slave which breaks the script completely (which some forum posts have confirmed this issue exists). This will also improve performance of Node Red on slower systems also (as loops are expensive even with latest Node), these fixes shouldn't be removed.**

4. TT: Updated config to use api.sunsynk.net and sunsynk.net for login, etc
GW: Unneeded
**TT: I have reverted**

5. TT: Reorder template as this broke in newer versions of HA 2024
GW: More information please. The latest version is running fine on the latest HA and Nodered
**TT: The issue was a missing `template:` at the top, as it was copied into a new HA installation, this was missing from the config, i've therefore added `template:`  to the top and fixed a typo - the wiki should be updated as this will catch people out.**

6. TT: Multiple lists of plants can have master and slaves and different size inverters, which broke part of the flow
GW: Raw data gather should be fine on this, so more info please
GW: Export flow doesnt currently cater for multi plant
**TT: See above comment for 3 as it is covered there**

@gdwaterworth please can you test the latest flow which is now updated

t0mtaylor commented:
Its worth noting this is currently looping through multi-plant, and with the fixes will still work for a single plant selected by the sn_data id and sn_flow id, the script will work as intended.

I will be looking at improving further to support multi-plant, but this may be sometime during February.

gdwaterworth commented:
PLease work in teh development branch, not the stable branch 
Also which flow nodes have you made changes on so I can do a comparison on the code

t0mtaylor commented:
- Setup Global Vars - new warning debug output mode
- Run Through Plant Inv List - additional debugging in debug mode
- Export Data Export To Entity - bug fixes as discussed in above comment for point 3

You should be able to run the flow after version 2.3 without issue

Let me know when your happy with the current changes, then I shall move the changes into the dev branch and test that and create a new PR into dev 👍

t0mtaylor commented:
@gdwaterworth are you happy with the changes?  I see you are doing a diff on the changes in the dev branch, which is good 👍 

Let me know and I will put them in dev branch and update the PR also

I have some more changes (which help with speeding up multi-plant setup and reduces error messages), but they will be in another branch and PR on dev once this one is merged

gdwaterworth commented:
Will look at the changes next week.

Regards

From: Tom Taylor ***@***.***>
Sent: Friday, February 2, 2024 8:20 PM
To: gdwaterworth/Home-Assistant---E-Linter-Logger---Sunsynk ***@***.***>
Cc: gdwaterworth ***@***.***>; Mention ***@***.***>
Subject: Re: [gdwaterworth/Home-Assistant---E-Linter-Logger---Sunsynk] Fixed bugs to support mixture of sunsynk inverter plant lists (PR #52)


@gdwaterworth<https://github.com/gdwaterworth> are you happy with the changes?

Let me know and I will put them in dev branch and update the PR also

I have some more changes, but they will be in another branch and PR on dev once this one is merged

-
Reply to this email directly, view it on GitHub<https://github.com/gdwaterworth/Home-Assistant---E-Linter-Logger---Sunsynk/pull/52#issuecomment-1924445579>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AUYVNI6QNBWDLW7IUOSGCXLYRUU6PAVCNFSM6AAAAABCTL27UWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMRUGQ2DKNJXHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

gdwaterworth commented:
OK been looking at a data issue and 0's and extra handling for sunsynk api timeout issues. Fixed and extra error checking and using filter sensor seem to fix/ mitigate that problem now 
Will start looking at yours. Just busy trying to get an account with master slave so I can look at the main issue.

gdwaterworth commented:
Implemented the changes in the export flow into the development branch

gdwaterworth commented:
The template change should not be required 

is allready at line ~417
I know the config file uses 2 different styles. Will convert all to one at some point

t0mtaylor commented:
Apologies for the delay @gdwaterworth, I've had issue with live data from sunsynk with my plants because they are stuck in region 1! Had this issue since January and they've only just agreed to get the engineering team to move the plants over to region2!

As soon as the migration has been completed to region2, I'll retest the latest dev branch and supply a new PR with any changes that may still be beneficial.
